subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -128,7 +128,7 @@ static int download_client_callback(const struct download_client_evt *event)
 						      dfu_target_callback_handler);
 				if ((err < 0) && (err != -EBUSY)) {
 					LOG_ERR("dfu_target_init error %d", err);
-					err_cause = FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED;
+					err_cause = FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE;
```

Comment:
I agree it should not be `_DOWNLOAD_FAILED`, since we actually managed to start the download, but I think we should add another error cause. In case of delta updates, we can't say the patch was not valid because we have failed before we managed to send forward it to the modem. Have not looked at the other kind of updates though.

Reply:
We are a bit limited on what we can express here:
```
/**
 * @brief FOTA download error cause values.
 */
enum fota_download_error_cause {
	/** No error, used when event ID is not FOTA_DOWNLOAD_EVT_ERROR. */
	FOTA_DOWNLOAD_ERROR_CAUSE_NO_ERROR,
	/** Downloading the update failed. The download may be retried. */
	FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED,
	/** The update is invalid and was rejected. Retry will not help. */
	FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE,
	/** Actual firmware type does not match expected. Retry will not help. */
	FOTA_DOWNLOAD_ERROR_CAUSE_TYPE_MISMATCH,
};
```

If the Delta update would be rejected, then I would assume the error code is `TYPE_MISMATCH`. It is unfortunate that `INVALID_UPDATE` sounds too similar. But that is the only error that fits into this.

Reply:
> If the Delta update would be rejected, then I would assume the error code is TYPE_MISMATCH

Yes, but the patch it's not being rejected. We can't say anything about the patch because we have failed at an earlier stage, during `dfu_target_init`, which in case of delta DFU can happen on:
- failure to read current modem firmware UUID
- failure to obtain size of scratch area
- failure to obtain scratch area offset

None of these have anything to do with the patch that has been downloaded so `_TYPE_MISMATCH` would not be a good fit according to its description. My suggestion was to add a new error to that list since none seem to apply to this case.
It seems all the other types of "targets" also do not fail on `_init` because of issues with the patch.
For example, a generic `_ERROR_CAUSE_INTERNAL`?

Reply:
I think the `ERROR_CAUSE_INTERNAL` makes sense.
I'll add that.

But then again, I really fail to see what is the difference between `TYPE_MISMATCH` and `INVALID_UPDATE`.
If you send garbage, is it INVALID or MISMATCH?
I would propose to drop that as well.

Reply:
I suppose that this library can download different types of firmware: modem, app, etc. Perhaps a different type of firmware was detected, not necessarily a bad image but an image the library was not expecting.